### PR TITLE
Update Function editor section (with Project functions functionality)

### LINKED ISSUE #9258: [feature] Allow users to save expression functions in QGIS project file (Request in QGIS)
### Request for documentation
From pull request qgis/QGIS#58354
Author: @gacarrillor
QGIS version: 3.40

**[feature] Allow users to save expression functions in QGIS project file**

### PR Description:
Current situation:
In the expression builder dialog, Python functions can only be saved in the user profile directory. If a user wants to share expression functions with colleagues, he/she needs to share his/her expressions directory or individual `.py` files, which is inconvenient (without mentioning the installation of those shared files won't be trivial either).

This PR allows users to also store expression functions into the current QGIS project file, making it easier to share functions with other interested parties.

Some details about the implementation:

  + At most, one `[Project Functions]` item will be listed in the left-hand side list widget (`Function Editor` tab).
  + Users will be able to add and remove the aforementioned item.
  + If a `[Project Functions]` item already exists in the list, the corresponding option in the `Add Function File` dialog will be disabled.
  + Project gets dirty when expression functions are added to the project or when they are edited.
  + When opening the QGIS project, we check user preferences about loading project functions (`Never|Ask|For this session only|Not during this session|Always`) and load project functions if they're allowed.
  + Project functions are unloaded when a project is closed, and the user functions are reloaded to avoid any potential overwrite done by project functions activated by a regular user (i.e., a user that does not alter Project functions but that loads (enabling them) them when opening the project).
    
    Note: This behavior differs with respect to expression functions saved in the user profile (i.e., when a function file is removed, expression functions are not unloaded for the ongoing QGIS session). The rationale behind this is that user functions need more knowledge from the user (she/he needs to write them or at least deal with Python files to install them), whereas in the case of projects, any (non-python-expert) user could trigger an overwrite just by opening the project with functions and allowing them in the QGIS session.

image

image

Unit tests included.

----------

Funded by the [QGIS user group Switzerland](https://qgis.ch/).

### Commits tagged with [need-docs] or [FEATURE]
"[feature] Allow to add/save/autosave/remove expression functions in a QGIS project"

### PR DESCRIPTION
**Pull request by gacarrillor**
Goal: Update Function editor section.

Ticket(s): #9258

- [ ] Backport to LTR documentation is requested

Fix #9258

### CODE REVIEW

File: docs/user_manual/introduction/qgis_configuration.rst [hunk 1]
```diff
@@ -132,8 +132,9 @@ displayed at the bottom of the frame.
   version of QGIS`. You can always open projects created with older version of
   QGIS but once the project is saved, trying to open with older release may fail
   because of features not available in that version.
-* :guilabel:`Enable macros` |selectString|. This option was created to handle
-  macros that are written to perform an action on project events. You can
+* :guilabel:`Enable project's embedded Python code` |selectString|. This option
+  was created to handle macros that are written to perform an action on project
```
Comment (DelazJ) [inline docs/user_manual/introduction/qgis_configuration.rst:136]: Let's add an anchor and remove the "was created"
```suggestion
.. _load_project_code:

* :guilabel:`Enable project's embedded Python code` |selectString|. This option
  handles execution of macros that are written to perform an action on project
```

**Review by DelazJ — COMMENTED** [review]
Thanks @gacarrillor

File: docs/user_manual/expressions/expression.rst [hunk 2]
```diff
@@ -335,12 +335,24 @@ particular needs that would not be covered by the predefined functions.
 To create a new function:
 
 #. Press the |symbologyAdd| :sup:`New File` button.
-#. Enter a name to use in the form that pops up and press :guilabel:`OK`.
+#. Select where to store the new function. You can choose to save it either in a
+   Function file or in the current QGIS project file.
```
[the remark this answers is not in the document]
Comment (gacarrillor) [reply docs/user_manual/expressions/expression.rst:339]: If you want to store your functions in the project file, there'll be only one item in the left panel for all of them, that's why the option is in plural.

File: docs/user_manual/expressions/expression.rst [hunk 3]
```diff
@@ -352,20 +364,29 @@ To create a new function:
...
    tab.
 
-Custom Python functions are stored under the user profile directory, meaning that at
-each QGIS startup, it will auto load all the functions defined with the current user
-profile. Be aware that new functions are only saved in the :file:`/python/expressions`
-folder and not in the project file.
+Custom Python functions stored as Function files are stored under the user profile
+directory, meaning that at each QGIS startup, it will auto load all the functions
+defined with the current user profile. Be aware that new functions are only saved in
+the :file:`/python/expressions` folder and not in the project file.
```
[the remark this answers is not in the document]
Comment (gacarrillor) [reply docs/user_manual/expressions/expression.rst:370]: Agreed, I rephrased it to avoid any confusion.

**Comment by gacarrillor** [comment]
Thanks for your suggestions @DelazJ

File: docs/user_manual/introduction/qgis_configuration.rst [hunk 4]
```diff
@@ -132,8 +132,11 @@ displayed at the bottom of the frame.
   version of QGIS`. You can always open projects created with older version of
   QGIS but once the project is saved, trying to open with older release may fail
   because of features not available in that version.
-* :guilabel:`Enable macros` |selectString|. This option was created to handle
-  macros that are written to perform an action on project events. You can
+.. _load_project_code:
```
Comment (DelazJ) [inline docs/user_manual/introduction/qgis_configuration.rst:135]:
```suggestion

.. _load_project_code:
```

**Pull request merged**